django/contrib/postgres/search.py

```diff
@@ -12,6 +14,16 @@
 from django.db.models.functions import Cast, Coalesce
 
 
+def normalize_spaces(val: str):
+    """Converts multiple spaces to single and strips from both sides."""
+    return re.sub(r"\s{2,}", " ", val.strip()) if val else None
+
+
+def psql_escape(query: str):
+    """Replace unsafe chars with space and convert multiple spaces to single."""
+    return normalize_spaces(re.sub(r"['\0\[\]()|&:*!@<>\\]", " ", query))
```

Comment:
Have you looked into how much of this you can delegate to the database adapter? I don't think we want to maintain our own implementation of sanitizing parameters.

When I had to do something similar in a django project, I used a combination of `AsIs` and `QuotedString` to get this done on psycopg2; not saying that's the (only) blessed way, but there should be some way (and we'd need a psycopg3 version too).

Reply:
For now, I can't find a way to combine `QuotedString`, `AsIs`, and a few other adapter methods to replace all the special characters in the regex. Another approach could be to use database-level sanitation like below:
```py
def sanitize_lexeme_query(connection, query: str):
    """
    Sanitizes a lexeme search query using PostgreSQL's built-in text search sanitation.

    This method uses to_tsquery to handle:
    - Removing special characters
    - Normalizing spaces
    """
    if not query:
        return None

    with connection.cursor() as cursor:
        cursor.execute("SELECT to_tsquery(%s) AS safe_query", (query,))
        result = cursor.fetchone()
        return result[0] if result else None
```
could you link me the way you handled a similar situation and used `AsIs` and `QuotedString` to do so @jacobtylerwalls ?

Reply:
Database-level sanitization looks intriguing. Thanks for raising it; we will likely want another opinion.

> could you link me the way you handled a similar situation 

Turns out my memory was of something substantially different, sorry.